File: app/lib/features/contacts/models/friend_models.dart
import 'package:app/features/auth/models/auth_user.dart';

enum FriendRequestStatus { pending, accepted, declined }

FriendRequestStatus friendRequestStatusFromString(String value) {
  switch (value.toLowerCase()) {
    case 'accepted':
      return FriendRequestStatus.accepted;
    case 'declined':
      return FriendRequestStatus.declined;
    case 'pending':
    default:
      return FriendRequestStatus.pending;
  }
}

String friendRequestStatusToString(FriendRequestStatus status) {
  switch (status) {
    case FriendRequestStatus.accepted:
      return 'accepted';
    case FriendRequestStatus.declined:
      return 'declined';
    case FriendRequestStatus.pending:
      return 'pending';
  }
}

enum FriendRequestAction { accept, decline }

class FriendInfo {
  const FriendInfo({
    required this.id,
    required this.user,
    required this.createdAt,
    this.remark,
  });

  final String id;
  final AuthUser user;
  final DateTime createdAt;
  final String? remark;

  String get displayName {
    final normalizedRemark = remark?.trim();
    return normalizedRemark?.isNotEmpty == true
        ? normalizedRemark!
        : user.displayName;
  }

  FriendInfo copyWith({String? remark, bool clearRemark = false}) {
    return FriendInfo(
      id: id,
      user: user,
      createdAt: createdAt,
      remark: clearRemark ? null : (remark ?? this.remark),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'user': user.toJson(),
    'created_at': createdAt.toIso8601String(),
    'friend_remark': remark,
  };

  factory FriendInfo.fromJson(Map<String, dynamic> json) {
    return FriendInfo(
      id: json['id'] as String? ?? '',
      user: AuthUser.fromJson(json['user'] as Map<String, dynamic>),
      createdAt: DateTime.parse(json['created_at'] as String),
      remark: json['friend_remark'] as String?,
    );
  }
}

class EnsureChatResult {
  const EnsureChatResult({
    required this.roomId,
    required this.roomName,
    required this.roomType,
    required this.friendId,
    required this.friendName,
    this.friendAvatar,
  });

  final String roomId;
  final String roomName;
  final String roomType;
  final String friendId;
  final String friendName;
  final String? friendAvatar;

  factory EnsureChatResult.fromJson(Map<String, dynamic> json) {
    return EnsureChatResult(
      roomId: json['room_id'] as String? ?? '',
      roomName: json['room_name'] as String? ?? '',
      roomType: json['room_type'] as String? ?? 'private',
      friendId: json['friend_id'] as String? ?? '',
      friendName: json['friend_name'] as String? ?? '',
      friendAvatar: json['friend_avatar'] as String?,
    );
  }
}

class FriendRequestInfo {
  const FriendRequestInfo({
    required this.id,
    required this.requester,
    required this.addressee,
    required this.status,
    required this.createdAt,
    this.respondedAt,
    this.message,
    this.isIncoming = false,
  });

  final String id;
  final AuthUser requester;
  final AuthUser addressee;
  final FriendRequestStatus status;
  final DateTime createdAt;
  final DateTime? respondedAt;
  final String? message;
  final bool isIncoming;

  AuthUser get counterparty => isIncoming ? requester : addressee;

  factory FriendRequestInfo.fromJson(Map<String, dynamic> json) {
    return FriendRequestInfo(
      id: json['id'] as String? ?? '',
      requester: AuthUser.fromJson(json['requester'] as Map<String, dynamic>),
      addressee: AuthUser.fromJson(json['addressee'] as Map<String, dynamic>),
      status: friendRequestStatusFromString(
        json['status'] as String? ?? 'pending',
      ),
      message: json['message'] as String?,
      isIncoming: json['is_incoming'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      respondedAt: json['responded_at'] != null
          ? DateTime.tryParse(json['responded_at'] as String)
          : null,
    );
  }
}

File: ios-app/Sources/RedCodeNetworking/FriendModels.swift
import Foundation
import RedCodeCore

public enum FriendRequestStatus: String, Codable, Equatable, Sendable {
    case pending
    case accepted
    case declined

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = (try? container.decode(String.self))?.lowercased()
        self = FriendRequestStatus(rawValue: rawValue ?? "") ?? .pending
    }
}

public enum FriendRequestAction: String, Codable, Equatable, Sendable {
    case accept
    case decline
}

public struct FriendInfo: Decodable, Equatable, Identifiable, Sendable {
    public let id: String
    public let user: AuthUser
    public let createdAt: Date?
    public let remark: String?

    public init(id: String, user: AuthUser, createdAt: Date? = nil, remark: String? = nil) {
        self.id = id
        self.user = user
        self.createdAt = createdAt
        self.remark = remark
    }

    public var displayName: String {
        if let remark, !remark.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return remark
        }
        if let nickname = user.nickname?.trimmingCharacters(in: .whitespacesAndNewlines), !nickname.isEmpty {
            return nickname
        }
        return user.username
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case user
        case createdAt = "created_at"
        case remark
        case friendRemark = "friend_remark"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? ""
        user = try container.decode(AuthUser.self, forKey: .user)
        createdAt = container.decodeFlexibleDate(forKey: .createdAt)
        remark = try container.decodeIfPresent(String.self, forKey: .friendRemark)
            ?? container.decodeIfPresent(String.self, forKey: .remark)
    }
}

public struct FriendRequestInfo: Decodable, Equatable, Identifiable, Sendable {
    public let id: String
    public let requester: AuthUser
    public let addressee: AuthUser
    public let status: FriendRequestStatus
    public let message: String?
    public let createdAt: Date?
    public let respondedAt: Date?
    public let isIncoming: Bool

    public init(
        id: String,
        requester: AuthUser,
        addressee: AuthUser,
        status: FriendRequestStatus,
        message: String? = nil,
        createdAt: Date? = nil,
        respondedAt: Date? = nil,
        isIncoming: Bool = false
    ) {
        self.id = id
        self.requester = requester
        self.addressee = addressee
        self.status = status
        self.message = message
        self.createdAt = createdAt
        self.respondedAt = respondedAt
        self.isIncoming = isIncoming
    }

    public var counterparty: AuthUser {
        isIncoming ? requester : addressee
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case requester
        case addressee
        case status
        case message
        case createdAt = "created_at"
        case respondedAt = "responded_at"
        case isIncoming = "is_incoming"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? ""
        requester = try container.decode(AuthUser.self, forKey: .requester)
        addressee = try container.decode(AuthUser.self, forKey: .addressee)
        status = try container.decodeIfPresent(FriendRequestStatus.self, forKey: .status) ?? .pending
        message = try container.decodeIfPresent(String.self, forKey: .message)
        createdAt = container.decodeFlexibleDate(forKey: .createdAt)
        respondedAt = container.decodeFlexibleDate(forKey: .respondedAt)
        isIncoming = try container.decodeIfPresent(Bool.self, forKey: .isIncoming) ?? false
    }
}

public struct EnsurePrivateChatResult: Decodable, Equatable, Sendable {
    public let roomID: String
    public let roomName: String
    public let roomType: ChatType
    public let friendID: String
    public let friendName: String
    public let friendAvatarURL: String?
    public let friendAvatarObjectKey: String?

    public init(
        roomID: String,
        roomName: String,
        roomType: ChatType = .privateChat,
        friendID: String,
        friendName: String,
        friendAvatarURL: String? = nil,
        friendAvatarObjectKey: String? = nil
    ) {
        self.roomID = roomID
        self.roomName = roomName
        self.roomType = roomType
        self.friendID = friendID
        self.friendName = friendName
        self.friendAvatarURL = friendAvatarURL
        self.friendAvatarObjectKey = friendAvatarObjectKey
    }

    private enum CodingKeys: String, CodingKey {
        case roomID = "room_id"
        case roomName = "room_name"
        case roomType = "room_type"
        case friendID = "friend_id"
        case friendName = "friend_name"
        case friendAvatarURL = "friend_avatar"
        case friendAvatarObjectKey = "friend_avatar_object_key"
    }
}

public struct CreateFriendRequestPayload: Encodable, Equatable, Sendable {
    public let targetUserID: String
    public let message: String?

    public init(targetUserID: String, message: String? = nil) {
        self.targetUserID = targetUserID.trimmingCharacters(in: .whitespacesAndNewlines)
        self.message = message?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
    }

    private enum CodingKeys: String, CodingKey {
        case targetUserID = "target_user_id"
        case message
    }
}

public struct RespondFriendRequestPayload: Encodable, Equatable, Sendable {
    public let action: FriendRequestAction

    public init(action: FriendRequestAction) {
        self.action = action
    }
}

private extension KeyedDecodingContainer {
    func decodeFlexibleDate(forKey key: Key) -> Date? {
        if let string = try? decodeIfPresent(String.self, forKey: key) {
            return FriendDateParser.parse(string)
        }
        if let double = try? decodeIfPresent(Double.self, forKey: key) {
            let seconds = double > 1_000_000_000_000 ? double / 1000 : double
            return Date(timeIntervalSince1970: seconds)
        }
        return nil
    }
}

private enum FriendDateParser {
    static func parse(_ value: String?) -> Date? {
        guard let value, !value.isEmpty else {
            return nil
        }

        let fractionalFormatter = ISO8601DateFormatter()
        fractionalFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = fractionalFormatter.date(from: value) {
            return date
        }

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}
